File: 30-substring-with-concatenation-of-all-words/30-substring-with-concatenation-of-all-words.java
class Solution {
    public List<Integer> findSubstring(String s, String[] words) {
        int m = s.length(), n = words.length, w = words[0].length();
        Map<String, Integer> tot = new HashMap<>();
        for(String word : words) tot.put(word, tot.getOrDefault(word, 0) + 1);

        List<Integer> res = new ArrayList<>();
        for(int i = 0; i < w; i++) {
            Map<String, Integer> window = new HashMap<>();
            int cnt = 0;
            for(int j = i; j + w <= m; j += w) {
                //如果窗口大小大约 n*w
                if(j >= i + n * w) {
                    String left = s.substring(j - n*w, j - (n-1)*w);
                    //System.out.println("left is " + left);
                    window.put(left, window.getOrDefault(left, 0) - 1);
                    if(window.get(left) < tot.getOrDefault(left, 0)) cnt--;
                }
                String right = s.substring(j, j + w);
                //.out.println("right is " + right);
                window.put(right, window.getOrDefault(right, 0) + 1);
                if(window.get(right) <= tot.getOrDefault(right, 0)) cnt++;
                if(cnt  == n) res.add(j - (n - 1) * w);
            }
        }
        return res;
    }
}

File: 30-substring-with-concatenation-of-all-words/30-substring-with-concatenation-of-all-words.py
class Solution:
    def findSubstring(self, s: str, words: List[str]) -> List[int]:
        l, n, w = len(s), len(words), len(words[0])
        counter = Counter(words)
        
        res = []
        for i in range(w):
            window = defaultdict(int)
            cnt = 0
            for j in range(i, l - w + 1, w):
                if j - i >= n * w:
                    left = s[j - n * w : j - (n - 1) * w]
                    window[left] -= 1
                    if window[left] < counter[left]:
                        cnt -= 1
                right = s[j : j + w]
                window[right] += 1
                if window[right] <= counter[right]:
                    cnt += 1
                if cnt == n:
                    res.append(j - (n - 1) * w)
        return res